File: src/bot/listeners/contents/documentHandler.ts
import { debug } from 'console';
import { Client } from 'tdl';
import { messageDocument, updateNewMessage } from 'tdlib-types';
import { DOCUMENT_DESTINATION } from '../../../constants';
import { isSender } from '../../filters';
import { refreshChats } from '../../helper';

export const documentHandler = async (
  update: updateNewMessage,
  client: Client,
  content: messageDocument
): Promise<void> => {
  // only accept the exact sender
  if (!isSender(update)) {
    debug('Received update from other sender', JSON.stringify(update));
    return;
  }

  // accepts only media files
  const mimeType = content.document.mime_type;
  if (!mimeType.startsWith('image') && !mimeType.startsWith('video')) {
    debug('Received wrong mime type', JSON.stringify(content));
    return;
  }

  console.log('Received new document', mimeType);

  // get chats
  const { topic } = await refreshChats(client, update.message.chat_id, {
    chatId: DOCUMENT_DESTINATION.chat,
    topicName: DOCUMENT_DESTINATION.thread,
  });

  // forward files
  await client.invoke({
    _: 'forwardMessages',
    chat_id: DOCUMENT_DESTINATION.chat,
    message_thread_id: topic?.info.message_thread_id,
    message_ids: [update.message.id],
    from_chat_id: update.message.chat_id,
  });
};

File: src/main/java/com/pirasalbe/services/telegram/listeners/contents/DocumentHandler.java
package com.pirasalbe.services.telegram.listeners.contents;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.pirasalbe.configurations.TelegramConfiguration;
import com.pirasalbe.helpers.FilterHelper;
import com.pirasalbe.helpers.TelegramBotHelper;
import com.pirasalbe.helpers.models.ForumInfo;
import com.pirasalbe.helpers.models.RefreshResult;

import it.tdlight.client.SimpleTelegramClient;
import it.tdlight.jni.TdApi.ForwardMessages;
import it.tdlight.jni.TdApi.MessageDocument;
import it.tdlight.jni.TdApi.UpdateNewMessage;

@Component
public class DocumentHandler {

	private static final Logger LOGGER = LoggerFactory.getLogger(DocumentHandler.class);

	@Autowired
	private TelegramConfiguration configuration;

	public void accept(SimpleTelegramClient client, UpdateNewMessage update, MessageDocument content) {
		if (!FilterHelper.isSender(configuration.getChatId(), update)) {
			LOGGER.debug("Received update from other sender", update);
			return;
		}

		// accepts only media files
		String mimeType = content.document.mimeType;
		if (!mimeType.startsWith("image") && !mimeType.startsWith("video")) {
			LOGGER.debug("Received wrong mime type {}", content);
			return;
		}

		LOGGER.info("Received new document {}", mimeType);

		// get chats
		RefreshResult refreshResult = TelegramBotHelper.refreshChats(client, update.message.chatId, new ForumInfo(
				configuration.getDocumentDestinationChatId(), configuration.getDocumentDestinationThreadName()));

		TelegramBotHelper.sendSync(client,
				new ForwardMessages(configuration.getDocumentDestinationChatId(),
						refreshResult.getTopic().map(topic -> topic.info.messageThreadId).orElse(0l),
						update.message.chatId, new long[] { update.message.id }, null, false, false));
	}

}
